File: Baekjoon/11000.js
const readline = require('readline');

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

let N = -1;
let classRoom = 1;
let schedule = [];

rl.on('line', (input) => {
    if (N == -1) {
        N = parseInt(input);
    } else {
        N -= 1;
        const [_s, _e] = input.split(' ').map(Number);
        schedule.push({start: _s, end: _e});
        if (N == 0) {
            rl.close();
        }
    }
});

rl.on('close', () => {
    schedule.sort((a, b) => {
        const [aLen, bLen] = [a.end - a.start, b.end - b.start];
        return bLen - aLen;
    })

    let arrange = {start: schedule[0].start, end: schedule[0].end};
    for (let i = 1 ; i < schedule.length ; i++) {
        if (arrange.start < schedule[i].start && arrange.end > schedule[i].end) {
            arrange.start = schedule[i].start;
            arrange.end = schedule[i].end;
            classRoom += 1;
        } else {
            if (schedule[i].start < arrange.end && arrange.end < schedule[i].end) {
                arrange.start = schedule[i].start;
                classRoom += 1;
            }
            else if (schedule[i].start < arrange.start && arrange.start < schedule[i].end) {
                arrange.end = schedule[i].end;
                classRoom += 1;
            }
        }
    }

    console.log(classRoom);
    process.exit();
});

File: Baekjoon/11000.py
import heapq
import sys

N = int(sys.stdin.readline())
inputArr = []

for _ in range(N):
    start, end = map(int, sys.stdin.readline().split())
    inputArr.append((start,end))

inputArr.sort()

endTimeHeap = []

heapq.heappush(endTimeHeap, inputArr[0][1])
for startT, endT in inputArr[1:]:
    if (endTimeHeap[0] > startT):
        heapq.heappush(endTimeHeap, endT)
    else:
        heapq.heappop(endTimeHeap)
        heapq.heappush(endTimeHeap, endT)

print(len(endTimeHeap))